### DISCUSSION TITLE
Log4j2 on Tomcat 10 fine with Console appender but only logs the startup Tomcat messages, not mine

### DISCUSSION THREAD

**Discussion by cnGitHubBruno** [apache/logging-log4j2#2994] [Q&A]
I am 99% certain that I messed up something in the log4j2-tomcat.xml but I can't see it.  Or is it something else?
I would greatly welcome any advice.

Syptoms:
- Log4j2 logs perfectly the Tomcat logs and all CONSOLE appender entries (I have it configured for DEBUG and even that shows fine).
- Only Tomcat startup messages show in my main.txt file from the MAIN appender.  Nothing from my code is showing in it... even though it shows fine in CONSOLE.

Configuration:
- Log4j version:  2.20.0 (yes, I plan to update to latest version before prod)
- Java version: 17
- Tomcat version: 10
- Files in %CATALINA_HOME%\log4j2\lib:   log4j-api.jar, log4j-appserver.jar, log4j-core.jar
- Location of log4j2-tomcat.xml:  %CATALINA_HOME%\log4j2\conf
- Commented original Tomcat %CATALINA_HOME\conf\logging.properties

log4j2-tomcat.xml file contents:

```<?xml version="1.0" encoding="utf-8"?>
<Configuration status="info">
  <Properties>
    <Property name="logdir">D:/appLogs</Property>
    <Property name="archivedir">${logdir}/archive</Property>
    <Property name="svrlogdir">${logdir}/server</Property>
    <Property name="svrarchivedir">${svrlogdir}/archive</Property>
    <Property name="layout">%d %-5p [%c:%L] - %m%n</Property>
  </Properties>
  <Appenders>
    <!-- CONSOLE: apps and "server leftovers" not yet moved to CATALINA -->
    <Console name="CONSOLE" target="SYSTEM_OUT">
      <PatternLayout pattern="${layout}"/>
    </Console>
    <!-- MAIN: apps and "server leftovers" not yet moved to CATALINA -->
    <RollingFile name="MAIN"
                 fileName="${logdir}/main.txt"
                 filePattern="${archivedir}/main.%d{yyyy-MM-dd}.txt.gz">
      <PatternLayout pattern="${layout}"/>
      <CronTriggeringPolicy schedule="0 0 0 * * ?"/>
      <DefaultRolloverStrategy>
        <Delete basePath="${archivedir}" maxDepth="1">
          <IfFileName glob="main.*.txt.gz" />
          <IfAccumulatedFileCount exceeds="10" />
        </Delete>
      </DefaultRolloverStrategy>
    <!-- alternative config, would replace CronTriggeringPolicy and DefaultRolloverStrategy
       <Policies>
       <TimeBasedTriggeringPolicy />
       <SizeBasedTriggeringPolicy size="10 MB"/>
      </Policies>
      <DefaultRolloverStrategy max="24"/>
    -->
    </RollingFile>
    <!-- SERVER appenders -->
    <RollingFile name="CATALINA"
                 fileName="${svrlogdir}/catalina.txt"
                 filePattern="${svrarchivedir}/catalina.%d{yyyy-MM-dd}.txt.gz">
      <PatternLayout pattern="${layout}"/>
      <CronTriggeringPolicy schedule="0 0 0 * * ?"/>
      <DefaultRolloverStrategy>
        <Delete basePath="${svrarchivedir}" maxDepth="1">
          <IfFileName glob="catalina.*.txt.gz" />
          <IfAccumulatedFileCount exceeds="10" />
        </Delete>
      </DefaultRolloverStrategy>
    </RollingFile>
    <RollingFile name="LOCALHOST"
                 fileName="${svrlogdir}/localhost.txt"
                 filePattern="${svrarchivedir}/localhost.%d{yyyy-MM-dd}.txt.gz">
      <PatternLayout pattern="${layout}"/>
      <CronTriggeringPolicy schedule="0 0 0 * * ?"/>
      <DefaultRolloverStrategy>
        <Delete basePath="${svrarchivedir}" maxDepth="1">
          <IfFileName glob="localhost.*.txt.gz" />
          <IfAccumulatedFileCount exceeds="10" />
        </Delete>
      </DefaultRolloverStrategy>
    </RollingFile>
    <RollingFile name="MANAGER"
                 fileName="${svrlogdir}/manager.txt"
                 filePattern="${svrarchivedir}/manager.%d{yyyy-MM-dd}.txt.gz">
      <PatternLayout pattern="${layout}"/>
      <CronTriggeringPolicy schedule="0 0 0 * * ?"/>
      <DefaultRolloverStrategy>
        <Delete basePath="${svrarchivedir}" maxDepth="1">
          <IfFileName glob="manager.*.txt.gz" />
          <IfAccumulatedFileCount exceeds="10" />
        </Delete>
      </DefaultRolloverStrategy>
    </RollingFile>
    <RollingFile name="HOST-MANAGER"
                 fileName="${svrlogdir}/host-manager.txt"
                 filePattern="${svrarchivedir}/host-manager.%d{yyyy-MM-dd}.txt.gz">
      <PatternLayout pattern="${layout}"/>
      <CronTriggeringPolicy schedule="0 0 0 * * ?"/>
      <DefaultRolloverStrategy>
        <Delete basePath="${svrarchivedir}" maxDepth="1">
          <IfFileName glob="host-manager.*.txt.gz" />
          <IfAccumulatedFileCount exceeds="10" />
        </Delete>
      </DefaultRolloverStrategy>
    </RollingFile>
  </Appenders>
  <Loggers>
    <!-- CNDC loggers -->
    <Root level="info">
      <AppenderRef ref="CONSOLE" />
      <AppenderRef ref="MAIN"/>
    </Root>
    <Logger name="net.cndc" level="debug" />
    <!-- OPEN SOURCE library loggers -->
    <Logger name="org.springframework" level="warn" />
    <!-- SERVER loggers -->
    <!--   for CATALINA appender -->
    <Logger name="org.apache.catalina.startup" level="info" additivity="false"><AppenderRef ref="CATALINA" /></Logger>
    <Logger name="org.apache.catalina.core" level="info" additivity="false"><AppenderRef ref="CATALINA" /></Logger>
    <Logger name="org.apache.coyote" level="info" additivity="false"><AppenderRef ref="CATALINA" /></Logger>
    <Logger name="org.apache.tomcat" level="info" additivity="false"><AppenderRef ref="CATALINA" /></Logger>
    <!-- -->
    <Logger name="org.apache.catalina.core.ContainerBase.[Catalina].[localhost]"
            level="info"
            additivity="false">
      <AppenderRef ref="LOCALHOST"/>
    </Logger>
    <Logger name="org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/manager]"
            level="info"
            additivity="false">
      <AppenderRef ref="MANAGER"/>
    </Logger>
    <Logger name="org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/host-manager]"
            level="info"
            additivity="false">
      <AppenderRef ref="HOST-MANAGER"/>
    </Logger>
  </Loggers>
</Configuration>
```

**Comment by rgoers** [comment]
I suggest you enable the StatusLogger debug logging. I am betting you will see Log4j initialize twice - once for Tomcat and once for your web app. I am also betting your web app will NOT be using log4j2-tomcat.xml.

**Comment by ppkarwasz** [comment]
Hi @cnGitHubBruno,

If your applications are bundled with `log4j-api`, due to the "parent-last" delegation model of web application classloaders, they will not be able to use your global Log4j API installation.

I recently published a [couple of Tomcat classloaders](https://oss.copernik.eu/tomcat/3.x/components/tomcat-log4j#classloaders) that allow administrators to use a global Log4j API installation without having to entirely invert the classloader delegation model.
In your case you just need to add [`tomcat-log4j-3.0.0-beta1.jar`](https://mvnrepository.com/artifact/eu.copernik/tomcat-log4j/3.0.0-beta1) to the `%CATALINA_HOME%\lib` directory (not `\log4j2\lib`) and add:

```xml
<!-- Use global Log4j API global installation -->
<Loader loaderClass="eu.copernik.tomcat.log4j.loader.Log4jParallelWebappClassLoader"/>
```

to your `%CATALINA_HOME%\conf\context.xml` file.
If you prefer to use the standard classloaders, you can always invert the delegation model:

```xml
<!-- Prefer global libraries over application specific -->
<Loader delegate="true"/>
```

Check also my [Logging in Tomcat using Log4j Core](https://oss.copernik.eu/tomcat/3.x/guide) guide.

**Note**: These artifacts are published under my personal `eu.copernik` namespace and they are not part of the official Apache Log4j distribution.
They will not be included in the Apache Log4j project until they have a decent user base.

**Reply by cnGitHubBruno** [reply]
As you guys mentioned, it seems likely that Tomcat inverted its priority on the log4j config that gets used sometime between Tomcat 8 and Tomcat 10.

Given that it seemed like the ideal solution would be using the Log4jParallelWebappClassLoader but it completely killed my apps.  Could it be that it is because my apps were written for and tomcat 10 configured with log4j 2.20.0 and not 3.0.0-beta?  (and I really would prefer not having to modify my whole codebase for 3.0 at this time)

If so... is there a 2.x version?

**Reply by ppkarwasz** [reply]
As mentioned before `tomcat-log4j` is a **third-party** artifact and is not part of the Apache Log4j distribution. Its version **does not** follow the versioning of Log4j2, so `tomcat-log4j` version `3.0.0-beta1` can be used with Apache Log4j `2.24.0`.

If you find a problem with its usage, please report it to [copernik-eu/log4j/issues](https://github.com/copernik-eu/log4j-tomcat/issues).

**Remark**: as far as I remember, Tomcat did no changes to its classloaders between version 8.5 and 10.1.
